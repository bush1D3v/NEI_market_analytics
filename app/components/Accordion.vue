<script lang="ts" setup>
import {
	Accordion,
	AccordionContent,
	AccordionItem,
	AccordionTrigger,
} from "@/components/ui/accordion";
import type {HTMLAttributes} from "vue";

export interface Item {
	value: string;
	title: string;
	content: string;
}

export interface AccordionProps {
	items: Item[];
	type?: "single" | "multiple" | undefined;
	collapsible?: boolean;
	class?: HTMLAttributes["class"];
}

const props = withDefaults(defineProps<AccordionProps>(), {
	type: "single",
	collapsible: true,
	class: "w-full",
});
</script>

<template>
    <Accordion :type="props.type" :class="props.class" :collapsible="props.collapsible" data-testid="Accordion">
        <AccordionItem v-for="item in props.items" :key="item.value" :value="item.value">
            <AccordionTrigger>
                <span v-translate>
                    {{ item.title }}
                </span>
            </AccordionTrigger>
            <AccordionContent>
                <span v-translate>
                    {{ item.content }}
                </span>
            </AccordionContent>
        </AccordionItem>
    </Accordion>
</template>
