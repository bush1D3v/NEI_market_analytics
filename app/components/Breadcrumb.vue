<script lang="ts" setup>
import {computed} from "vue";
import {useRoute, useRouter} from "vue-router";
import RouterLink from "@/tags/RouterLink.vue";
import {
	Breadcrumb,
	BreadcrumbItem,
	BreadcrumbLink,
	BreadcrumbList,
	BreadcrumbPage,
	BreadcrumbSeparator,
} from "@/components/ui/breadcrumb";

const route = useRoute();
const router = useRouter();

const breadcrumbs = computed(() => {
	const paths = route.path.split("/").filter((path) => path);
	const breadcrumbArray = paths.map((path, index) => {
		const to = `/${paths.slice(0, index + 1).join("/")}`;
		const routeMatch = router.resolve(to);
		return {
			text: routeMatch.name || path,
			to,
		};
	});
	return [{text: "Home", to: "/"}, ...breadcrumbArray];
});
</script>

<template>
    <Breadcrumb class="pt-4 pl-4 lg:pb-4">
        <BreadcrumbList>
            <BreadcrumbItem v-for="(breadcrumb, index) in breadcrumbs" :key="index">
                <BreadcrumbLink v-if="index < breadcrumbs.length - 1">
                    <RouterLink :to="breadcrumb.to">{{ breadcrumb.text }}</RouterLink>
                </BreadcrumbLink>
                <BreadcrumbPage v-else>{{ breadcrumb.text }}</BreadcrumbPage>
                <BreadcrumbSeparator v-if="index < breadcrumbs.length - 1" />
            </BreadcrumbItem>
        </BreadcrumbList>
    </Breadcrumb>
</template>
