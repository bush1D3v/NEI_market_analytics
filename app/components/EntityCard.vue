<script setup lang="ts">
import Image from "@/tags/Image.vue";
import Button from "@/components/ui/button/Button.vue";
import numberFormatter from "@/utils/numberFormatter";

export interface EntityCardProps {
	id: string | number;
	name: string;
	symbol: string;
	image: string;
	routerLinkTo: string;
	price: number;
	market_cap: number;
	circulating_supply: number;
}

const props = defineProps<EntityCardProps>();
</script>

<template>
    <article class="w-80 flex flex-col bg-secondaryDarker gap-4 p-4 md:p-5 rounded-lg md:w-96">
        <div class="flex gap-4 items-center">
            <Image :src="props.image" :alt="props.name + ' logo'" width="50" height="50"
                class="w-10 h-10 md:w-[50px] md:h-[50px] rounded-full" />
            <div class="flex flex-col">
                <h3 class="line-clamp-1 max-w-56">{{ props.name }}</h3>
                <div class="flex gap-1 md:gap-[6px]">
                    <p>Currency</p>
                    <strong>{{ props.symbol.toUpperCase() }}</strong>
                    <p>&nbsp;·&nbsp; Chain ID</p>
                    <strong class="max-w-8 md:max-w-12 line-clamp-1">{{ props.id }}</strong>
                </div>
            </div>
        </div>
        <hr class="h-[1px] w-full" />
        <h6 v-translate>Estatísticas</h6>
        <div class="md:font-semibold gap-2 md:gap-0 flex justify-between">
            <div>
                <strong>${{ numberFormatter(props.price) }}</strong>
                <p v-translate>Preço</p>
            </div>
            <div>
                <strong>{{ numberFormatter(props.market_cap) }}</strong>
                <p v-translate>Capitalização de Mercado</p>
            </div>
            <div>
                <strong>{{ numberFormatter(props.circulating_supply) }}</strong>
                <p v-translate>Volume</p>
            </div>
        </div>
        <RouterLink :to="props.routerLinkTo">
            <Button class="w-full">Explore {{ props.name }}</Button>
        </RouterLink>
    </article>
</template>
