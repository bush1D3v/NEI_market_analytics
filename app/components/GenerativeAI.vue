<script lang="ts" setup>
import generateContent, {type GenerateContentProps} from "@/services/Gemini";
import socket from "@/server/Socket";
import Button from "@/components/ui/button/Button.vue";
import Image from "@/tags/Image.vue";
import RouterLink from "@/tags/RouterLink.vue";
import LoadingDots from "./Loading/ChatDots.vue";
import LikeImage from "@/assets/images/like.png";
import DislikeImage from "@/assets/images/dislike.png";
import useToastNotification from "@/notification/toast";
import {v4 as uuidv4} from "uuid";
import {ref, nextTick} from "vue";
import {
	Bot,
	User,
	BotMessageSquare,
	Send,
	RefreshCw,
	Copy,
	ThumbsDown,
	ThumbsUp,
} from "lucide-vue-next";
import {Collapsible, CollapsibleContent, CollapsibleTrigger} from "@/components/ui/collapsible";

interface Message {
	text: string;
	sender: "ai" | "user";
	refs: string[];
	first?: boolean;
	isLiked?: boolean;
	isDisliked?: boolean;
}

const defaultMessage: Message = {
	text: "ðŸ‘‹ Hi! I am NEI Market AI, ask me anything about NEI Market Analytics!",
	sender: "ai",
	refs: [],
	first: true,
};

const prompt = ref("");
const messages = ref<Message[]>([defaultMessage]);
const loading = ref(false);
const error = ref(false);
const sessionId = ref(uuidv4());
const isOpen = ref(false);

let aiMessage = "";

function copyToClipboard(text: string) {
	navigator.clipboard
		.writeText(text)
		.then(() => {
			useToastNotification("Text copied to clipboard");
		})
		.catch((_e) => {
			useToastNotification("Error copying text to clipboard");
		});
}

function toggleLike(message: Message) {
	message.isLiked = !message.isLiked;
	if (message.isLiked) {
		message.isDisliked = false;
	}
}

function toggleDislike(message: Message) {
	message.isDisliked = !message.isDisliked;
	if (message.isDisliked) {
		message.isLiked = false;
	}
}

function refresh() {
	sessionId.value = uuidv4();
	aiMessage = "";
	messages.value = [defaultMessage];
	useToastNotification("Chat refreshed");
}

async function sendMessage(resendMessage?: boolean) {
	loading.value = true;

	const lastUserMessage = [...messages.value]
		.reverse()
		.find((message) => message.sender === "user");

	const userMessage = resendMessage ? (lastUserMessage?.text as string) : prompt.value;
	if (userMessage.trim() === "" && !resendMessage) return;

	const lastBotMessage = [...messages.value].reverse().find((message) => message.sender === "ai");

	if (resendMessage) {
		messages.value.pop();
	} else {
		messages.value.push({text: userMessage, sender: "user", refs: []});
		nextTick(() => {
			const messagesContainer = document.querySelector(".chat-messages");
			if (messagesContainer) {
				messagesContainer.scrollTop = messagesContainer.scrollHeight;
			}
		});
	}
	prompt.value = "";

	try {
		const data: GenerateContentProps = {
			prompt: userMessage,
			sessionId: sessionId.value,
			lastUserMessage: lastUserMessage?.text,
			lastBotMessage: lastBotMessage?.text,
			likeOrDislikePreviousMessage: lastBotMessage?.isLiked
				? true
				: lastBotMessage?.isDisliked
					? false
					: undefined,
		};
		await generateContent(data);
	} catch (error) {
		console.error(error);
	}
}

function adjustTextareaHeight(event: Event) {
	const textarea = event.target as HTMLTextAreaElement;
	textarea.style.height = "auto";
	textarea.style.height = `${textarea.scrollHeight}px`;
}

socket.on("content", (data) => {
	// biome-ignore lint/suspicious/noDoubleEquals: <explanation>
	if (data.sessionId == sessionId.value) {
		aiMessage += data.text;
		if (
			messages.value.length > 0 &&
			messages.value[messages.value.length - 1].sender === "ai"
		) {
			messages.value[messages.value.length - 1].text = aiMessage;
		} else {
			messages.value.push({text: aiMessage, sender: "ai", refs: data.refs});
		}
		loading.value = false;
	}
});

socket.on("content_end", (requestSessionId) => {
	// biome-ignore lint/suspicious/noDoubleEquals: <explanation>
	if (requestSessionId == sessionId.value) {
		aiMessage = "";
		nextTick(() => {
			const messagesContainer = document.querySelector(".chat-messages");
			if (messagesContainer) {
				messagesContainer.scrollTop = messagesContainer.scrollHeight;
			}
		});
	}
});

socket.on("error", (data) => {
	// biome-ignore lint/suspicious/noDoubleEquals: <explanation>
	if (data.sessionId == sessionId.value) {
		error.value = true;
		loading.value = false;
		console.error(data.error);
	}
});
</script>

<template>
    <Collapsible v-model:open="isOpen" class="fixed right-0 lg:right-14 bottom-32 lg:bottom-14 z-50">
        <CollapsibleTrigger as-child class="absolute right-4 lg:-right-10 -bottom-10">
            <Button variant="default" class="w-12 h-12 p-2 rounded-full">
                <BotMessageSquare :size="50" />
            </Button>
        </CollapsibleTrigger>
        <CollapsibleContent>
            <div class="chat-container  lg:w-[400px]">
                <header>
                    <div class="flex items-center gap-2">
                        <Image src="/logo-just-icon.png" alt="NEI Market AI Logo" class="brightness-[100]" width="40"
                            height="40" />
                        <h6>NEI Market AI</h6>
                    </div>
                    <button title="Recarregar chat" @click="refresh" :disabled="messages.length === 1 || loading"
                        class="disabled:opacity-50 hover:opacity-50 duration-150 ease-in-out">
                        <RefreshCw :size="20" :stroke-width="3" />
                    </button>
                </header>
                <ul class="chat-messages" ref="messagesContainer">
                    <li v-for="(message, index) in messages" :key="index">
                        <div :class="[ 'chat-message', message.sender ]" class="rounded-full relative">
                            <Bot v-if="message.sender === 'ai'" :class="[ 'chat-icon', message.sender ]" />
                            <User v-if="message.sender === 'user'" :class="[ 'chat-icon', message.sender ]" />
                            <span class="text-white">{{ message.text }}</span>
                        </div>
                        <div v-if="message.sender === 'ai' && !message.first"
                            class="chat-actions ml-7 -mt-1 flex gap-3">
                            <button title="Copiar" @click="copyToClipboard(message.text)">
                                <Copy :size="16" />
                            </button>
                            <button title="Like" @click="toggleLike(message)">
                                <ThumbsUp :size="16" v-if="!message.isLiked" />
                                <Image :src="LikeImage" alt="Like" width="16" height="16" class="dark:invert" v-else />
                            </button>
                            <button title="Dislike" @click="toggleDislike(message)">
                                <ThumbsDown :size="16" v-if="!message.isDisliked" />
                                <Image :src="DislikeImage" alt="Dislike" width="16" height="16" class="dark:invert"
                                    v-else />
                            </button>
                            <button v-if="index === messages.length - 1" title="Recarregar" @click="sendMessage(true)">
                                <RefreshCw :size="16" />
                            </button>
                        </div>
                    </li>
                    <LoadingDots v-if="loading && !error" />
                    <p v-if="error" class="text-negative">Erro ao gerar conteÃºdo, tente novamente mais tarde!</p>
                </ul>
                <div class="chat-input">
                    <textarea class="flex w-full" maxlength="1000" v-model="prompt" @input="adjustTextareaHeight"
                        @keyup.enter="sendMessage()" placeholder="Digite sua mensagem..." />
                    <Button title="NEI Market AI Chat"
                        class="p-2 absolute right-4 bottom-[10px] dark:hover:bg-black/25 hover:bg-slate-100"
                        variant="ghost" @click="() => { error = false, sendMessage() }"
                        :disabled="loading || prompt.trim() === ''">
                        <Send :size="20" color="var(--text)" :stroke-width="2.5" />
                    </Button>
                </div>
                <footer>
                    <p class="text-text text-sm">By chatting, you agree to our</p>
                    <RouterLink class="underline text-sm hover:opacity-50" to="/privacy-policy">privacy policy
                    </RouterLink>
                    .
                </footer>
            </div>
        </CollapsibleContent>
    </Collapsible>
</template>

<style scoped lang="css">
.chat-container {
    height: 80dvh;
    margin: 0 auto;
    border: 1px solid var(--border);
    border-radius: 8px;
    display: flex;
    flex-direction: column;
    margin-bottom: 10px;
    background-color: var(--background);
}

header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: 10px;
    padding: 10px;
    background-color: var(--primary);
    color: #fff;
    text-align: center;
    border-top-left-radius: 8px;
    border-top-right-radius: 8px;
}

.chat-messages {
    flex: 1;
    padding: 10px;
    overflow-y: auto;
    display: flex;
    flex-direction: column;
    gap: 16px;
}

.chat-messages div {
    position: relative;
    display: flex;
    align-items: center;
}

.chat-icon {
    position: absolute;
    width: 20px;
    height: 20px;
    bottom: 0px
}

.chat-messages div .chat-icon.ai {
    left: -25px;
}

.chat-messages div .chat-icon.user {
    right: -25px;
}

.chat-message {
    margin-bottom: 10px;
    padding: 8px 16px;
    width: 85%;
}

.chat-message.user {
    background-color: var(--secondary);
    margin-left: auto;
    margin-right: 20px;
    border-radius: 20px 20px 1px 20px;
}

.chat-message.ai {
    background-color: var(--primary);
    align-self: flex-start;
    margin-right: auto;
    margin-left: 20px;
    border-radius: 20px 20px 20px 1px;
}

.chat-actions svg,
.chat-actions div {
    cursor: pointer;
}

.chat-actions svg:hover,
.chat-actions div:hover {
    opacity: 50%;
    transition-duration: 150ms;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
}

.chat-input {
    display: flex;
    position: relative;
    align-items: center;
    border-radius: 1px;
}

.chat-input textarea {
    color: var(--text);
    background: var(--secondary-darker);
    max-height: 200px;
    padding: 4px;
    padding-right: 60px;
    resize: none;
    box-shadow: 0 0.5px 0px var(--text), 0 -0.5px 0px var(--text);
}

.chat-input textarea:focus {
    outline: none;
    box-shadow: 0 0.5px 1px var(--text), 0 -0.5px 0px var(--text);
}

.chat-input textarea::placeholder {
    padding-top: 12px;
}

footer {
    display: flex;
    align-content: center;
    justify-content: center;
    padding: 4px;
    gap: 4px;
}
</style>
