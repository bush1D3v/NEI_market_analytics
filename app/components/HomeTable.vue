<script setup lang="ts">
import Image from "@/tags/Image.vue";
import RouterLink from "@/tags/RouterLink.vue";
import numberFormatter from "@/utils/numberFormatter";
import type {CryptoCurrency} from "@/types/CoinMarketCap/CryptoCurrency";
import type {Stock} from "@/types/BrapiDev/Stock";
import type {New} from "@/types/Finnhub/New";
import Link from "@/tags/Link.vue";

interface Props {
	icon: string;
	tableTitle: string;
	redirectTo: "/cryptos" | "/news" | "/stocks";
	coins?: CryptoCurrency[];
	stocks?: Stock[];
	news?: New[];
}

const props = defineProps<Props>();
</script>

<template>
    <article>
        <div class="flex justify-between">
            <div class="flex gap-1 items-center">
                <span> {{ props.icon }} </span>
                <h4> {{ props.tableTitle }} </h4>
            </div>
            <RouterLink :to="props.redirectTo" target="_self" class="text-link hover:underline mt-2 h-fit">
                See more
            </RouterLink>
        </div>
        <hr class="h-[1px] w-full bg-gray-300" />
        <ul class="flex flex-col gap-2">
            <li v-if="props.coins" v-for="(data, i) in props.coins" :key="i">
                <div class="li-first-container">
                    <div class="flex gap-2">
                        <Image class="rounded-full"
                            :src="`https://s2.coinmarketcap.com/static/img/coins/64x64/${data.id}.png`"
                            :alt="data.name + ' logo'" width="24" height="24" />
                        <RouterLink :to="`/cryptos/${data.name}`" target="_self" class="hover:underline">
                            <h5 class="line-clamp-1 text-left"> {{ data.name }} </h5>
                            </ RouterLink>
                    </div>
                    <span class="hidden mn:flex text-dark"> {{ data.symbol }} </span>
                </div>
                <div class="li-second-container">
                    <span> {{ numberFormatter(data.quote.USD.price) }} </span>
                    <span class="min-w-12 text-left"> {{ numberFormatter(data.quote.USD.volume_change_24h) }} </span>
                </div>
            </li>
            <li v-if="props.stocks" v-for="(data, i) in props.stocks" :key="i">
                <div class="li-first-container">
                    <div class="flex gap-2">
                        <Image class="rounded-full" :src="data.logo" :alt="data.name + ' logo'" width="24"
                            height="24" />
                        <RouterLink :to="`/stocks/${data.name}`" target="_self" class="hover:underline">
                            <h5 class="line-clamp-1 text-left"> {{ data.name }} </h5>
                            </ RouterLink>
                    </div>
                    <span class="hidden mn:flex text-dark"> {{ data.stock }} </span>
                </div>
                <div class="li-second-container">
                    <span> {{ numberFormatter(data.close) }} </span>
                    <span class="min-w-12 text-left"> {{ numberFormatter(data.change) }} </span>
                </div>
            </li>
            <li v-if="props.news" v-for="(data, i) in props.news" :key="i" class="flex gap-3">
                <Image :src="data.image" :alt="data.headline + ' logo'" width="82" height="47"
                    class="w-[82px] h-[47px] rounded-lg" />
                <div class="flex flex-col max-h-14 text-left">
                    <Link :href="data.url" target="_blank" class="hover:underline">
                    <h6 class="line-clamp-2 max-w-56 min-w-56 md:max-w-72 md:min-w-72 text-sm">
                        {{ data.headline }}
                    </h6>
                    </Link>
                    <span class="text-xs text-dark">{{ new Date(data.datetime * 1000).toLocaleDateString('en-US', {
                        month: 'short',
                        day: 'numeric'
                    }) }}</span>
                </div>
            </li>
        </ul>
    </article>
</template>

<style lang="css" scoped>
article {
    @apply flex flex-col justify-around md:justify-center bg-secondaryDarker rounded-lg gap-1 p-2 md:p-3 min-w-[310px] max-w-[310px] mn:min-w-[355px] mn:max-w-[355px] min-h-[250px] max-h-[250px] md:min-w-[425px] md:max-w-[425px] md:min-h-[250px] md:max-h-[250px];
}

.li-first-container {
    @apply flex gap-2 min-w-40 max-w-40 mn:min-w-60 mn:max-w-60 md:min-w-72 md:max-w-72 justify-between;
}

.li-second-container {
    @apply flex gap-2;
}

li {
    @apply flex gap-2 items-center justify-around mn:justify-between text-center;
}

span {
    @apply text-sm md:text-base;
}
</style>
