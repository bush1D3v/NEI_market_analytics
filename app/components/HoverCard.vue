<script setup lang="ts">
import {CalendarIcon} from "@radix-icons/vue";

import {Button} from "@/components/ui/button";
import {HoverCard, HoverCardContent, HoverCardTrigger} from "@/components/ui/hover-card";
import Image from "@/tags/Image.vue";

export interface HoverCardProps {
	title: string;
	description: string;
	startDate: string;
	avatarSrc: string;
}

const props = defineProps<HoverCardProps>();
</script>

<template>
    <HoverCard>
        <HoverCardTrigger as-child>
            <Button variant="link" class="p-0 justify-start w-fit" data-testid="HoverCardButton">
                <span>@{{ props.title }}</span>
            </Button>
        </HoverCardTrigger>
        <HoverCardContent class="w-80">
            <div class="flex justify-between space-x-4">
                <Image :src="props.avatarSrc" :alt="props.title" width="80" height="80" class="rounded-full" />
                <div class="space-y-1">
                    <h4 class="text-sm font-semibold">
                        @{{ props.title }}
                    </h4>
                    <p class="text-sm">
                        {{ props.description }}
                    </p>
                    <div class="flex items-center pt-2">
                        <CalendarIcon class="mr-2 h-4 w-4 opacity-70" />
                        <span class="text-xs text-textForeground">
                            Joined {{ props.startDate }}
                        </span>
                    </div>
                </div>
            </div>
        </HoverCardContent>
    </HoverCard>
</template>
