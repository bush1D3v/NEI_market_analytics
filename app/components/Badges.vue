<script lang="ts" setup>
import Image from "@/tags/Image.vue";
import Link from "@/tags/Link.vue";
import HoverCard from "@/components/HoverCard.vue";

export interface BadgeProps {
	name: string;
	occupation: string;
	github: string;
	nickname: string;
	description: string;
	startDate: string;
}

const props = defineProps<{badges: BadgeProps[]}>();
</script>

<template>
    <ul class="flex gap-4 flex-wrap justify-center" data-testid="Badges">
        <li v-for="badge in props.badges" class="flex gap-2 min-w-64">
            <div>
                <Link :href="badge.github" target="_blank">
                <Image :src="badge.github + '.png'" :alt="badge.name + ' photo'" width="80" height="80"
                    class="rounded-full" />
                </Link>
            </div>
            <div class="flex flex-col">
                <div class="flex flex-col gap-y-0">
                    <h3>{{ badge.name }}</h3>
                    <span class="ml-[2px] -mt-[6px]">{{ badge.occupation }}</span>
                </div>
                <HoverCard :description="badge.description" :title="badge.nickname" :start-date="badge.startDate"
                    :avatar-src="badge.github + '.png'" />
            </div>
        </li>
    </ul>
</template>
