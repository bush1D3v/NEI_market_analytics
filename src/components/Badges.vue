<script lang="ts">
import Image from "@/tags/Image.vue";
import {BadgesDto} from "./Dto/BadgesDto";
import Skeleton from "@/components/ui/skeleton/Skeleton.vue";
import Link from "@/tags/Link.vue";
import HoverCard from "@/components/HoverCard.vue";
import ImageSkeletonSvg from "@/components/Svg/ImageSkeletonSvg.vue";
import type {PropType} from "vue";

export interface BadgeProps {
	name: string;
	occupation: string;
	github: string;
	nickname: string;
	description: string;
	startDate: string;
	initialLetters: string;
}

export default {
	props: {
		badges: {
			type: Array as PropType<BadgeProps[]>,
			default: () => BadgesDto,
		},
	},
	data() {
		return {
			loading: true,
		};
	},
	components: {Skeleton, Image, Link, HoverCard, ImageSkeletonSvg},
	methods: {
		onImageLoad() {
			this.loading = false;
		},
	},
};
</script>

<template>
    <ul class="flex gap-4 flex-wrap justify-center" data-testid="Badges">
        <li v-for="badge in $props.badges" class="flex gap-2 min-w-64">
            <div>
                <Skeleton v-if="loading" class="w-20 h-20 rounded-full">
                    <ImageSkeletonSvg />
                </Skeleton>
                <Link :href="badge.github" target="_blank" class="hover:opacity-50 duration-150 ease-in-out">
                <Image :src="badge.github + '.png'" :alt="badge.name + ' photo'" width="80" height="80"
                    class="rounded-full" @load="onImageLoad()" />
                </Link>
            </div>
            <div class="flex flex-col">
                <div class="flex flex-col gap-y-0">
                    <h3>{{ badge.name }}</h3>
                    <span class="ml-[2px] -mt-[6px]">{{ badge.occupation }}</span>
                </div>
                <HoverCard :description="badge.description" :title="badge.nickname" :start-date="badge.startDate"
                    :avatar-src="badge.github + '.png'" :avatar-fallback="badge.initialLetters" />
            </div>
        </li>
    </ul>
</template>
