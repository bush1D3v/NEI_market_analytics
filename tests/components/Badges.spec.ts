import {test, expect, describe} from "vitest";
import Badges from "@/components/Badges.vue";
import {mount} from "@vue/test-utils";
import {BadgesMock} from "../../mocks/components/Badges";

describe("Badges Component Tests", () => {
	const wrapper = mount(Badges, {props: {badges: BadgesMock}});
	const badgesComponent = wrapper.findComponent(Badges);

	test("Should be able to render the component correctly", async () => {
		expect(badgesComponent).toBeDefined();
	});

	test("Should be able to snapshot the component correctly", async () => {
		expect(badgesComponent.html()).toMatchSnapshot();
	});

	test("Should be able to render the component with the correct text", async () => {
		for (const item of BadgesMock) {
			expect(badgesComponent.text()).toContain(item.name);
			expect(badgesComponent.text()).toContain(item.nickname);
			expect(badgesComponent.text()).toContain(item.occupation);
		}
	});

	test("Should be able to render the component with the correct classes", async () => {
		expect(badgesComponent.classes()).toStrictEqual([
			"flex",
			"gap-4",
			"flex-wrap",
			"justify-center",
		]);
	});
});
