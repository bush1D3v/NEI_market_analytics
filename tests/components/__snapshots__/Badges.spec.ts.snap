// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badges Component Tests > Should be able to snapshot the component correctly 1`] = `
"<ul class="flex gap-4 flex-wrap justify-center" data-testid="Badges">
  <li class="flex gap-2 min-w-64">
    <div><a href="https://github.com/bush1D3v" target="_blank" rel="no-referrer" class="hover:opacity-50 duration-150 ease-in-out" data-testid="Link">
        <div class="relative w-[80px] h-[80px]">
          <div class="animate-pulse bg-surface flex justify-center items-center w-full h-full absolute rounded-full"><svg class="w-8 h-8" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 18" role="img" focusable="false" aria-labelledby="svgTitle svgDesc">
              <title id="svgTitle">Image Placeholder</title>
              <desc id="svgDesc">A placeholder image in the shape of a rectangle with a mountain icon.</desc>
              <path d="M18 0H2a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2Zm-5.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Zm4.376 10.481A1 1 0 0 1 16 15H4a1 1 0 0 1-.895-1.447l3.5-7A1 1 0 0 1 7.468 6a.965.965 0 0 1 .9.5l2.775 4.757 1.546-1.887a1 1 0 0 1 1.618.1l2.541 4a1 1 0 0 1 .028 1.011Z"></path>
            </svg></div><img src="https://github.com/bush1D3v.png" alt="Victor Navarro photo" width="80" height="80" class="rounded-full" data-testid="Image">
        </div>
      </a></div>
    <div class="flex flex-col">
      <div class="flex flex-col gap-y-0">
        <h3>Victor Navarro</h3><span class="ml-[2px] -mt-[6px]">Software Engineer</span>
      </div><button class="inline-flex items-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 text-primary underline-offset-4 hover:underline duration-300 h-9 p-0 justify-start w-fit" data-testid="HoverCardButton" data-state="closed" data-grace-area-trigger=""><span>@bush1D3v</span> </button>
      <!--teleport start-->
      <!--teleport end-->
    </div>
  </li>
  <li class="flex gap-2 min-w-64">
    <div><a href="https://github.com/eracksonsouza" target="_blank" rel="no-referrer" class="hover:opacity-50 duration-150 ease-in-out" data-testid="Link">
        <div class="relative w-[80px] h-[80px]">
          <div class="animate-pulse bg-surface flex justify-center items-center w-full h-full absolute rounded-full"><svg class="w-8 h-8" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 18" role="img" focusable="false" aria-labelledby="svgTitle svgDesc">
              <title id="svgTitle">Image Placeholder</title>
              <desc id="svgDesc">A placeholder image in the shape of a rectangle with a mountain icon.</desc>
              <path d="M18 0H2a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2Zm-5.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Zm4.376 10.481A1 1 0 0 1 16 15H4a1 1 0 0 1-.895-1.447l3.5-7A1 1 0 0 1 7.468 6a.965.965 0 0 1 .9.5l2.775 4.757 1.546-1.887a1 1 0 0 1 1.618.1l2.541 4a1 1 0 0 1 .028 1.011Z"></path>
            </svg></div><img src="https://github.com/eracksonsouza.png" alt="Erackson Souza photo" width="80" height="80" class="rounded-full" data-testid="Image">
        </div>
      </a></div>
    <div class="flex flex-col">
      <div class="flex flex-col gap-y-0">
        <h3>Erackson Souza</h3><span class="ml-[2px] -mt-[6px]">Frontend Developer</span>
      </div><button class="inline-flex items-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 text-primary underline-offset-4 hover:underline duration-300 h-9 p-0 justify-start w-fit" data-testid="HoverCardButton" data-state="closed" data-grace-area-trigger=""><span>@eracksonsouza</span> </button>
      <!--teleport start-->
      <!--teleport end-->
    </div>
  </li>
  <li class="flex gap-2 min-w-64">
    <div><a href="https://github.com/IsraelPina32" target="_blank" rel="no-referrer" class="hover:opacity-50 duration-150 ease-in-out" data-testid="Link">
        <div class="relative w-[80px] h-[80px]">
          <div class="animate-pulse bg-surface flex justify-center items-center w-full h-full absolute rounded-full"><svg class="w-8 h-8" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 18" role="img" focusable="false" aria-labelledby="svgTitle svgDesc">
              <title id="svgTitle">Image Placeholder</title>
              <desc id="svgDesc">A placeholder image in the shape of a rectangle with a mountain icon.</desc>
              <path d="M18 0H2a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2Zm-5.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Zm4.376 10.481A1 1 0 0 1 16 15H4a1 1 0 0 1-.895-1.447l3.5-7A1 1 0 0 1 7.468 6a.965.965 0 0 1 .9.5l2.775 4.757 1.546-1.887a1 1 0 0 1 1.618.1l2.541 4a1 1 0 0 1 .028 1.011Z"></path>
            </svg></div><img src="https://github.com/IsraelPina32.png" alt="Israel Pina photo" width="80" height="80" class="rounded-full" data-testid="Image">
        </div>
      </a></div>
    <div class="flex flex-col">
      <div class="flex flex-col gap-y-0">
        <h3>Israel Pina</h3><span class="ml-[2px] -mt-[6px]">FullStack Developer</span>
      </div><button class="inline-flex items-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 text-primary underline-offset-4 hover:underline duration-300 h-9 p-0 justify-start w-fit" data-testid="HoverCardButton" data-state="closed" data-grace-area-trigger=""><span>@IsraelPina32</span> </button>
      <!--teleport start-->
      <!--teleport end-->
    </div>
  </li>
</ul>"
`;
